### PR TITLE
Make Tabs Great [Again] [*]

### PR DESCRIPTION
**Pull request by shundhammer**
[*] _No, they were never great in libyui-qt to begin with..._

# Trello

https://trello.com/c/WaSOYCD3/4683-make-qt-tabs-looks-great-again

# Bugzilla

https://bugzilla.opensuse.org/show_bug.cgi?id=1186705

# GitHub Issue Tracker

https://github.com/libyui/libyui/issues/20

# Problems

- Tabs in libyui-qt stretch over the complete width of the tab bar. This is unusual and ugly.
  See also https://github.com/libyui/libyui/issues/20 (includes screenshots).

- There are few visual cues: While the tabs at the top give some indication that this is meant as an analogy to cardboard filing cards, that look and feel completely evaporates in the lower part. There is no frame, no subtle different background color, nothing.

  The overall visual appearance leaves to be desired.

# Fixes

- For the width: Explicitly set the QTabBar's _expanded_ property to _false_. That's the easy part.

- For the visual appearance: Now no longer only using a QTabBar, but the real thing: A QTabWidget.

# Cheating our Way into the QTabWidget

## Challenges

The QTabWidget does everything all by itself; it is a pretty self-sufficient widget. The application adds tab _pages_, and the QTabWidget does the rest: It creates and populates a QTabBar with the page names, it manages the pages in a QWidgetStack, it puts a page into the foreground when the user clicks on a tab. All widgets for all pages coexist the whole time.

Our YDumbTab on the other hand is, as the name implies, _dumb_. Very dumb, to be exact. It has YItems, one for each tab; and when the user clicks on the tab, the application is notified about that with a YMenuEvent with the name of the new tab. The application can then choose to replace the content with something new; that's why a YDumbTab usually has a YReplacePoint as its only child (it can only have one). When receiving such a YMenuEvent, the application exchanges the page with `UI.ReplaceWidget()` with a new page.

This also means that the widgets only exist for as long as that page is visible; when the user switches to another page, the application has to fetch the value from those widgets and apply those values, then replace the widgets and fill the new widgets with values. This is how YaST applications using that YDumbTab widget are structured.

## Cheats Part 1: The Widget Tree and Layouts

Qt and its subclasses of QObject / QWidget manage widgets in a tree. Each parent widget owns its children and takes care to delete them in the destructor; and, equally important, some parent widget classes manage the _layout_ of their children. QTabBar is among them. Those widgets ask their children for their `sizeHint()`, add some spacings and margins and frames and whatnot, and add up the result.

Our libyui with its YWidget does the same thing; that is the basis of having a common abstract toolkit on top of NCurses, Qt and Gtk: Each manager widget asks its children their `preferredWidth()` and `preferredHeight()`, adds them up, compensates for too much or too little screen space, and assigns a final size (`setSize()`) to them to match all layout constraints.

And here we have two problems: You can't have two layout managers fighting each other. Normally, we don't use any of Qt's layout mechanisms in our libyui; we just use plain QWidgets for layout purposes, calculate the desired coordinates in our layout engine, and set fixed pixel coordinates and widget sizes. That's why the layouts are the same for the text-based NCurses and the graphical Qt and Gtk UI back-ends.

QTabWidget, however, has layout logic of its own, and that conflicts with our libyui layout logic. It has a QTabBar widget at the top that holds just the tabs and a QStackedWidget below which holds the pages with the content. And it does the layout management across all the pages to make sure there is enough screen space for all of them (in case the user switches to that page).

So what we are doing here with this pull request is giving the QTabBar _one_ tab with one page, and that's it. Only for the first YItem a QWidget is created and added as a page with the YItem's _label_ has the tab name. That satisfies the QTabWidget's requirements.

That `_firstPage` is also used to reparent the YWidget children that the application will create by reimplementing the `ChildEvent()` and checking for subclasses of `YWidget` and reparenting those to `_firstPage` so the resulting widget subtree looks like this:

```
    QTabWidget (YQDumbTab)
    ├── QTabBar
    └── QStackedWidget
        └── _firstPage (QWidget)
            └── YQReplacePoint
                └── YQVBox (page content)
```

_libyui-qt uses double inheritance; each widget class is subclassed from a QWidget subclass and from a YWidget subclass._

## Cheats Part 2: Switching Tab Pages

While that gives us a perfectly working _first_ tab page, that does not solve the conflict between QTabBar automatically switching to other pages (if it has any) and YDumbTab still needing to just send a YMenuEvent when the user clicks on a tab.

That problem is solved by simply not adding any more pages; it's always only the first one, and that first one (typically) contains a Y(Q)ReplacePoint that can be used by the application to cut out the existing widget subtree from there on and replace it with a new one.

But we have access to the QTabBar, and we liberally use that: We add _tabs_ for all the other YItems to that QTabBar and simply don't tell the QTabWidget about them. For all the QTabWidget knows, there is only one page, our `_firstPage`, and that's it. When it receives a Qt signal that the user clicked on another tab, it will simply not find a corresponding page (because the tab index is out of range) and do nothing.

But we also receive that Qt signal, and we send a YMenuEvent with the name of that tab to the application, and the application can handle it; it can replace the page content, it can ask the user first if any pending changes should be applied or abandoned, or whatever.

## The Nitty-Gritty: Margins, 3D borders, Spacings, other QStyle Oddities

One problem that is solved in only a pragmatic way and not really in general is that each QStyle may choose to use a different set of margins around each subwidget (the QTabBar or the QStackedWidget) or between them. It may choose to draw a 3D border around it, it may choose to add some spacing between them.

In the Qt library code of QTabWidget, there are orgies of `if (...)` cascades to figure out in what case to add what amount of margin and spacing, both in the `sizeHint()` that calculates how many pixels are needed, and in the actual paint event that draws the widget.

It turns out that in real life, it always uses 2 pixels of margin in both dimensions, and that's it. That was tested with different QStyles and QSS style sheets, and it always worked - so far. That does not mean that in some pathological case it might not choose to use 0 or 3 or 5 or whatever; from our perspective, this is not a pixel-perfect science. In real world applications, it shouldn't matter anyway; in the worst case, some content will get 2 or 3 pixels smaller or larger horizontally or vertically or both.

## Alternatives: Manually Drawing Widget Subcontrols via QStyle

For a while I considered to keep it the old way with only a QTabBar and a self-managed QWidget below and draw the borders manually in `QWidget::paintEvent()`. QStyle offers some dozen draw functions to draw parts of widgets ("subcontrols"). But those functions all want some subclass of QStyleOption with all kinds of sizes and dimensions and flags and whatnot; and it's crazy complicated to get those QStyleOptions initialized correctly. Adding insult to injury, the `initStyleOption()` functions for the interesting one (in particular the one needed here for QTabBar) tend to be _protected_, i.e. you can only use them from a subclass of that class (QTabBar in this case) which is exactly the one thing I didn't want.

If I have to inherit QTabBar, I might as well use its functionality to the extent that I need; and that's what I did.

## Cheating even More: Using Qt's Introspection to Access Subwidgets

For a while the code needed access to the QStackedWidget subwidget of QTabWidget, but that one is not exposed via the normal API: There is a getter method `QTabWidget::tabBar()` that returns the QTabBar, but no counterpart to get the QStackedWidget.

But... Qt can traverse its widget tree, and you can even search it for widgets of a certain type or a certain name; in this case, it was

```C++
QStackedWidget * stackedWidget = findChild<QStackedWidget *>();
```

...and hey presto, you have a pointer to the QTabBar's QStackedWidget, and you can use it in all its glory.

It turned out that this was not necessary after a few iterations of code change, but this is a method to get access to subwidgets that the Qt authors didn't want you to get your fingers on.

Is that an evil thing to to? Hell, yes!

Do we do that anyway? If we have to, of course!

So why do we do it if we actually shouldn't from the clean OOP point of view? **Because we can!** ;-)

# Screenshots

## Before

widget-demo-tw-03a

widget-demo-tw-03c

-----------------------------

## After

new-tabs-widget-demo-3a

new-tabs-widget-demo-3c

new-tabs-yast2-bootloader-1

new-tabs-yast2-bootloader-2

new-tabs-yast2-bootloader-3

new-tabs-yast2-lan-01

new-tabs-yast2-lan-02

new-tabs-yast2-users-old-1

### CODE REVIEW

**Review by hellcp — APPROVED** [review]

File: libyui-qt/src/YQDumbTab.cc [hunk 1]
```diff
@@ -175,18 +240,18 @@ YQDumbTab::preferredHeight()
...
-    return tabBarHeight + YQDumbTabSpacing + childHeight;
+    return tabBarHeight + childHeight;
 }
 
 
 void
 YQDumbTab::setSize( int newWidth, int newHeight )
 {
-    QWidget::resize( newWidth, newHeight );
+    // yuiDebug() << "new size: " << QSize( newWidth, newHeight ) << endl;
```
Comment (lslezak) [inline libyui-qt/src/YQDumbTab.cc:250]: NP: I'd either uncomment it (it's a debug only so it won't pollute the log in most cases) or just delete the line, I do not like dead code... :wink:

File: libyui-qt/src/YQDumbTab.cc [hunk 2]
```diff
@@ -228,19 +283,17 @@ YQDumbTab::setSize( int newWidth, int newHeight )
 	// Client area
 	//
 
-
+        // yuiDebug() << "client: " << QSize( remainingWidth, remainingHeight ) << "\n" << endl;
```
Comment (lslezak) [inline libyui-qt/src/YQDumbTab.cc:286]: The same here...

File: libyui-qt/src/YQUI.cc [hunk 3]
```diff
@@ -649,11 +649,15 @@ qMessageHandler( QtMsgType type, const QMessageLogContext &, const QString & msg
 #endif
 
 	case QtWarningMsg:
+            // Suppress well-known Qt warning that we can't do anything about
+            if ( msg.contains( "Timers cannot be stopped from another thread" ) )
```
Comment (lslezak) [inline libyui-qt/src/YQUI.cc:653]: :+1:

**Review by lslezak — APPROVED** [review]
NP: Just some commented debug logging.

I like the detailed PR description which contains a lot of interesting details but I'm afraid that it will be difficult to find later.

What about putting these details into a document (in the `doc/` subdirectory) or a [wiki page](https://github.com/libyui/libyui/wiki) (if it does not make sense in the code/package)?

File: libyui-qt/src/YQDumbTab.cc (hunk 1 above)
Comment (shundhammer) [reply libyui-qt/src/YQDumbTab.cc:250]: No, I intentionally left that to be easily reactivated when needed, yet not clutter the y2log all the time.

**Comment by shundhammer** [comment]
Added a link to this PR to the source file where it's most likely to be found. See PR #36.

**Pull request merged**